### DISCUSSION TITLE
[v4] Cannot apply unknown utility class

### DISCUSSION THREAD

**Discussion by ev3rlost** [tailwindlabs/tailwindcss#13336] [Help] [answered]
I have a seperate utilities.css file containing @import from other subfolders ('form/buttons.css', 'form/inputs.css', etc).

I keep getting this error:
```
[plugin:vite:css] [postcss] Cannot apply unknown utility class: flex-center
```

app.css
```
@import "tailwindcss";

@import 'fonts.css' layer(base);
@import 'base.css' layer(base);
@import 'utilities.css' layer(utilities);
@import 'components.css' layer(components);

@theme {
    --font-family-roboto: "Roboto", "sans-serif";
    --color-primary-50: #f5f3ff;
    --color-primary-100: #ede9fe;
    --color-primary-200: #ddd6fe;
    --color-primary-300: #c4b5fd;
    --color-primary-400: #a78bfa;
    --color-primary-500: #8b5cf6;
    --color-primary-600: #7c3aed;
    --color-primary-700: #6d28d9;
    --color-primary-800: #5b21b6;
    --color-primary-900: #4c1d95;
    --color-primary-950: #2e1065;
}
```

utilities.css
```
.flex-start { @apply flex justify-start items-center; }
.flex-center { @apply flex justify-center items-center; }
.flex-between { @apply flex justify-between items-center; }
.flex-end { @apply flex justify-end items-center; }
```

components.css
```
@import "form/buttons.css";
@import "form/inputs.css";
@import "form/radios.css";
@import "form/checkboxes.css";
@import "form/textareas.css";
@import "form/wysiwyg.css";
```

Components usualy use my custom utility classes for example flex-start

The CSS folder structure:
```
/css
   /form
      buttons.css
      inputs.css
      radios.css
      checkboxes.css
      textareas.css
      wysiwyg.css
   app.css
   base.css
   components.css
   fonts.css
   utilities.css
```

What am I missing?

**Comment by ev3rlost** [comment]
Bump! Anyone got any ideas what's causing the error?

**Comment by adamwathan** [comment, marked as answer]
Sounds like you're trying to do `@apply flex-center` somewhere — v4 doesn't currently support applying custom classes, only classes that are built in to the framework, so this just won't work quite yet. Will be in there eventually though before the real v4 release 👍

**Reply by ev3rlost** [reply]
Thank you Adam for clarifying! The hype is unreal! Can't wait for v4! :)

**Reply by bdrtsky** [reply]
Still not there. Just had this issue.

**Reply by mhenrixon** [reply]
How far away is this apply functionality, ya think @adamwathan? I am considering whether or not to start upgrading, and this would be inconvenient.

I love v4 so far btw, I am blown away by the speed and the new way of working.

**Reply by adamwathan** [reply]
As long as your custom utilities are registered as utilities using `@utility` it works today 👍🏻

**Reply by okcha4ever** [reply]
i still have this issue

**Reply by quedicesebas** [reply]
Me too

**Reply by vincent-dante** [reply]
I'm still getting error.

**Reply by Fasunle** [reply]
Any class that is throwing an error should be redeclared using **@utility** as shown below:

![image](https://github.com/user-attachments/assets/6316c9a9-96ea-465a-a035-fe516a6c81b1)

**Reply by mhenrixon** [reply]
Exactly what @Fasunle said.

I have had to put the utility definitions in the main file for my projects. 

I can't seem to add a @utility definition unless it is top-level. 

I have never looked into it; I just wanted to mention it in case someone struggles with the same.

**Reply by Fasunle** [reply]
Yeah... @utility cannot be nested. @mhenrixon 
 it has to be top-level declaration

**Reply by vincent-dante** [reply]
Thanks @Fasunle it works now.

First I have to make it all @utility

![CleanShot 2025-01-26 at 23 30 09@2x](https://github.com/user-attachments/assets/7924cce0-faff-4fe6-ba7e-ea34762040e0)

And now i can use it on my other class.
![CleanShot 2025-01-26 at 23 30 34@2x](https://github.com/user-attachments/assets/d243b1d6-2bd2-4134-ad54-2f7f55b4544a)

**Reply by vvp3** [reply]
I am using laravel and vue and I am trying to put tw 4 to work. I have in extended the color navyBlue and previosuly it worked in tailwind.config.js

extend: {
            colors: {
                navyBlue: "#060C3B",

Now I get the same error with unknown utility class. I tried also to use this in tailwind.config.js

plugins: [
            function ({ addUtilities }) {
                addUtilities({
                    '.bg-navyBlue': {
                        backgroundColor: '#060C3B',

In app.css @utility does not work and I tried something like this with no result though (i reckon is not tw native or I am not referencing something right?)

`
@layer components {

    --color-navyBlue: #060C3B;
    .bg-navyBlue {
        background-color: var(--color-navyBlue);
    }
    .border-navyBlue {
        border-color: var(--color-navyBlue);
    }

    .primary-button {
        @apply flex gap-x-1.5 items-center place-content-center font-medium rounded-xl text-white bg-navyBlue border border-navyBlue px-8 py-4 max-w-max cursor-pointer transition-all hover:opacity-[0.9];
    }
`

any help/ ideea would be great. thanks

**Reply by IRediTOTO** [reply]
This is not ready but you still release v4.0??? It make my site broken :(

**Reply by YutoYasunaga** [reply]
This release doesn't feel ready, and pushing it as a stable version makes it seem like you don't care about existing users. 

We love TailwindCSS and are grateful to the creator, but v4.0 has been a frustrating experience.

**Comment by illkle** [comment]
For anyone looking after upgrading to v4:

If you have a main css file where you have `@import "tailwindcss";` and some other css file with `@apply`, you will get an error unless you add `@reference "./global.css";`(file with tailwind import) to the top of that file.

**Reply by brennanangel** [reply]
Thank you for the tip!

If you keep getting `Error: Can't resolve 'theme.css` in .... I've found that the path name lookups are inconsistent depending on which TailwindCSS tool you're using. Try  `"./assets/css/theme.css";` vs. `"./theme.css"`, etc.

**Reply by sebastianvitterso** [reply]
Why is this not documented in the migration guide? This, as well as the change to using `@utility` should definitely be part of the migration guide.

**Reply by Verde19** [reply]
Thank you, it worked! Do you know if it's possible to do this all at once for components with their own <style> tags such as in Sveltekit? The different references to the global CSS file are a bit annoying.

**Reply by Verde19** [reply]
> Thank you, it worked! Do you know if it's possible to do this all at once for components with their own <style> tags such as in Sveltekit? The different references to the global CSS file are a bit annoying.

Solved it. As per [Tailwind's Sveltekit install guide](https://tailwindcss.com/docs/installation/framework-guides/sveltekit), use:

@reference "tailwindcss/theme";

instead of

@reference "[relative-path]/app.css";

**Reply by scottbedard** [reply]
Thanks for sharing this, here is more context from the docs

https://tailwindcss.com/docs/upgrade-guide#using-apply-with-vue-svelte-or-css-modules

**Reply by loxK** [reply]
> If you have a main css file where you have `@import "tailwindcss";` and some other css file with `@apply`, you will get an error unless you add `@reference "./global.css";`(file with tailwind import) to the top of that file.

I try to use it with Laravel. In the `<head />` of my main template I have `@vite(['resources/css/app.css', 'resources/js/app.js'])`. 

`app.css` :
```css
@import "tailwindcss";
@plugin "tailwindcss-primeui";
@import "./primevue/tailwind.css"; /* styles downloaded from the zip file 
```

This gives this error : `[plugin:@tailwindcss/vite:generate:serve] Cannot apply unknown utility class: shadow-[inset_0_2px_0_0_theme(colors.primary)]`

Where and how should I use `@reference`, please ?

**Reply by Abouhassane** [reply]
@loxK I have the same issue, have you found a solution ?

**Reply by loxK** [reply]
In the meantime, I noticed that tailwind support for PrimeVue will be abandoned soon. So I switched to Vuetify for that new project. PrimeVue updates always have a lot of breaking changes. Too much time consuming.

**Reply by Aqu1tain** [reply]
Thanks, solved the issue for me. 
with nuxt : 
```css
@reference "@/assets/css/main.css"; 
```
(Where my `@import "tailwindcss";` is)

**Comment by JaquesBotha** [comment]
I'm quite shocked at the fact that they push Tailwind 4 on us with such integral changes yet not document certain parts (IE:@apply) clearly. 

Don't force core changes on us if you not 110% ready. I've tried everything from trying to include the old tailwind.config file within css to @reference within the css file, **NOTHING** is working when it comes to @apply. Clearly not thought over properly.

Here's an example of what I'm refering to:
```
@import "tailwindcss" prefix(tw);

.quickgrid {
	@apply tw-table-auto
    tw-w-full;
}
/* Table Header */
.quickgrid thead th {
	@apply tw-bg-slate-100
    dark:tw-bg-slate-800/70
    tw-py-4;
}
```

If sometone knows of a solution please comment

**Reply by wongjn** [reply]
Prefixes in v4 are now separated by `:` and go at the beginning, so your CSS should look like:
```diff
 @import "tailwindcss" prefix(tw);

 .quickgrid {
- 	@apply tw-table-auto
-     tw-w-full;
+ 	@apply tw:table-auto
+     tw:w-full;
 }
 /* Table Header */
 .quickgrid thead th {
- 	@apply tw-bg-slate-100
-     dark:tw-bg-slate-800/70
-     tw-py-4;
+ 	@apply tw:bg-slate-100
+     tw:dark:bg-slate-800/70
+     tw:py-4;
 }
```
---
As an aside, Adam Wathan (creator of Tailwind) does seem to advocate avoiding `@apply`:
- https://twitter.com/adamwathan/status/1226511611592085504
- https://twitter.com/adamwathan/status/1559250403547652097

**Reply by JaquesBotha** [reply]
Do you have proof that the prefix has changed, I dont think it has? I dont see prefix documentation anywhere again to my point of not properly testing everying. I've tested your prefix change and still having the same errors.

**Reply by wongjn** [reply]
> Do you have proof that the prefix has changed, I dont think it has?

It was documented in the v4 beta docs: https://v3.tailwindcss.com/docs/v4-beta#using-a-prefix

> I've tested your prefix change and still having the same errors.

Seems to work in [this Tailwind Play](https://play.tailwindcss.com/iQGB564ydk?file=css) at least, so it could be something particular to your project. You may also need to use [the `@reference` directive](https://tailwindcss.com/docs/functions-and-directives#reference-directive) in places. Consider providing a remotely-hosted git repository if you'd like assistance with further debugging.

**Reply by Jyothi0612** [reply]
Hi,

After upgrading to Tailwind CSS v4, I'm encountering errors like "[plugin:vite:css] [postcss] Cannot apply unknown utility class: text-gray-200," despite using @reference.

 my app.css file looks like below
```
@import 'primeicons/primeicons.css';
@import './base/variables.css';
@import './base/darkmode.css';

@import './primevue/datatable.css';
@import './primevue/progressbar.css';
@import './primevue/input.css';
@import './primevue/dialog.css';
@import './primevue/toolbar.css';
@import './primevue/multiselect.css';
@import './primevue/buttons.css';
@import './primevue/actionbuttons.css';

@layer tailwind-base, primevue, tailwind-utilities;

@plugin '@tailwindcss/forms';
@plugin 'tailwindcss-primeui';

@source '../../vendor/laravel/framework/src/Illuminate/Pagination/resources/views/*.blade.php';

@custom-variant dark (&:is(.dark *));

@theme {
  --color-primary-blue: var(--primary-blue);
  --color-primary-blue-hover: var(--primary-blue-hover);
  --color-primary-blue-active: var(--primary-blue-active);
  --color-header-blue: var(--header-blue);
  --color-success: var(--success-color);
  --color-danger: var(--danger-color);

  --background-color-success-hover: var(--success-hover-bg);
  --background-color-danger-hover: var(--danger-hover-bg);

  --spacing-action-button: var(--action-button-size);
  --spacing-action-padding: var(--action-button-padding);
}

/*
  The default border color has changed to `currentColor` in Tailwind CSS v4,
  so we've added these compatibility styles to make sure everything still
  looks the same as it did with Tailwind CSS v3.

  If we ever want to remove these styles, we need to add an explicit border
  color utility to any element that depends on these defaults.
*/
@layer base {
  *,
  ::after,
  ::before,
  ::backdrop,
  ::file-selector-button {
    border-color: var(--color-gray-200, currentColor);
  }
}

@layer tailwind-base {
    @import 'tailwindcss';
}

@layer tailwind-utilities {
}

/* Header */
.header-title {
    @apply bg-[var(--header-blue)] text-white px-6 py-4 rounded-xl flex justify-between items-center;
    h1 {
        @apply text-lg font-normal;
    }
}

.page-container {
     @apply p-8 bg-gray-50 dark:bg-[#1a1f2c];
}
```

In my datatable.css, I have
```
@reference path="../app.css";
/* Column Headers */
    .p-datatable-thead > tr > th {
        @apply bg-[#f8fafc] dark:bg-[#1a1f2c]
               text-gray-700 dark:text-gray-200
               border-b border-gray-200 dark:border-0
               px-6 py-3;

        .p-column-header-content {
            @apply text-gray-700 dark:text-gray-200;
        }

        .p-column-filter {
            @apply mt-2;

            input {
                @apply bg-white dark:bg-[#1a1f2c]
                       border border-gray-200 dark:border-[#2f3744]
                       text-gray-700 dark:text-gray-200;
            }
        }

        .p-column-filter-menu-button {
            @apply text-gray-400 dark:text-gray-500 
                   hover:bg-gray-100 dark:hover:bg-[#1a1f2c]
                   rounded-lg;

            &.p-column-filter-menu-button-active {
                @apply bg-blue-50 dark:bg-blue-900/30
                       text-blue-500 dark:text-blue-400;
            }
        }
    }
```
[plugin:vite:css] [postcss] Cannot apply unknown utility class: text-gray-200

**Reply by JaquesBotha** [reply]
Yup I'm having the exact same issue. Clearly it's not ready yet to be upgrades to 4.0! Also, using the excuse not to use @apply and use the CSS Classes instead is not a solution when working with large projecs that make use of dynamic components...

**Comment by Jyothi0612** [comment]
Hi,

After upgrading to Tailwind CSS v4, I'm encountering errors like "[plugin:vite:css] [postcss] Cannot apply unknown utility class: text-gray-200," despite using @reference.

my app.css file looks like below
```
@import 'primeicons/primeicons.css';
@import './base/variables.css';
@import './base/darkmode.css';

@import './primevue/datatable.css';
@import './primevue/progressbar.css';
@import './primevue/input.css';
@import './primevue/dialog.css';
@import './primevue/toolbar.css';
@import './primevue/multiselect.css';
@import './primevue/buttons.css';
@import './primevue/actionbuttons.css';

@layer tailwind-base, primevue, tailwind-utilities;

@plugin '@tailwindcss/forms';
@plugin 'tailwindcss-primeui';

@source '../../vendor/laravel/framework/src/Illuminate/Pagination/resources/views/*.blade.php';

@custom-variant dark (&:is(.dark *));

@theme {
  --color-primary-blue: var(--primary-blue);
  --color-primary-blue-hover: var(--primary-blue-hover);
  --color-primary-blue-active: var(--primary-blue-active);
  --color-header-blue: var(--header-blue);
  --color-success: var(--success-color);
  --color-danger: var(--danger-color);

  --background-color-success-hover: var(--success-hover-bg);
  --background-color-danger-hover: var(--danger-hover-bg);

  --spacing-action-button: var(--action-button-size);
  --spacing-action-padding: var(--action-button-padding);
}

/*
  The default border color has changed to `currentColor` in Tailwind CSS v4,
  so we've added these compatibility styles to make sure everything still
  looks the same as it did with Tailwind CSS v3.

  If we ever want to remove these styles, we need to add an explicit border
  color utility to any element that depends on these defaults.
*/
@layer base {
  *,
  ::after,
  ::before,
  ::backdrop,
  ::file-selector-button {
    border-color: var(--color-gray-200, currentColor);
  }
}

@layer tailwind-base {
    @import 'tailwindcss';
}

@layer tailwind-utilities {
}

/* Header */
.header-title {
    @apply bg-[var(--header-blue)] text-white px-6 py-4 rounded-xl flex justify-between items-center;
    h1 {
        @apply text-lg font-normal;
    }
}

.page-container {
     @apply p-8 bg-gray-50 dark:bg-[#1a1f2c];
}
```
In my datatable.css, I have
```
@reference path="../app.css";
/* Column Headers */
    .p-datatable-thead > tr > th {
        @apply bg-[#f8fafc] dark:bg-[#1a1f2c]
               text-gray-700 dark:text-gray-200
               border-b border-gray-200 dark:border-0
               px-6 py-3;

        .p-column-header-content {
            @apply text-gray-700 dark:text-gray-200;
        }

        .p-column-filter {
            @apply mt-2;

            input {
                @apply bg-white dark:bg-[#1a1f2c]
                       border border-gray-200 dark:border-[#2f3744]
                       text-gray-700 dark:text-gray-200;
            }
        }

        .p-column-filter-menu-button {
            @apply text-gray-400 dark:text-gray-500 
                   hover:bg-gray-100 dark:hover:bg-[#1a1f2c]
                   rounded-lg;

            &.p-column-filter-menu-button-active {
                @apply bg-blue-50 dark:bg-blue-900/30
                       text-blue-500 dark:text-blue-400;
            }
        }
    }
    ```
[plugin:vite:css] [postcss] Cannot apply unknown utility class: text-gray-200

**Reply by JaquesBotha** [reply]
Already responded to this above but just to add Documentation and exact steps is essential for this kind of stuff.

**Comment by alexchexes** [comment]
Without `@reference` this works **perfectly** (Vue SFC):
```vue
<style lang="postcss">
.test-pre {
  @apply break-words whitespace-pre-wrap;
}
</style>
```

but this ↓ throws error `[vite] Internal server error: [postcss] Cannot apply unknown utility class: bg-white`:

```vue
<style lang="postcss">
.test-color {
  @apply bg-white;
}
</style>
```
What is the difference?

**Reply by wongjn** [reply]
See https://github.com/tailwindlabs/tailwindcss/issues/15778#issuecomment-2616412050 and https://github.com/tailwindlabs/tailwindcss/issues/16314#issuecomment-2641089872, but `break-words whitespace-pre-wrap` don't rely on a config, whereas `bg-white` does, hence the error.

---
As an aside, Adam Wathan (creator of Tailwind) does seem to advocate avoiding `@apply`:
- https://twitter.com/adamwathan/status/1226511611592085504
- https://twitter.com/adamwathan/status/1559250403547652097

You could consider using Tailwind class names in your template, instead of `@apply`.

**Reply by alexchexes** [reply]
> You could consider using Tailwind class names in your template, instead of `@apply`.

I need to style an external component (in this case, [AceEditor for Vue](https://github.com/CarterLi/vue3-ace-editor)), I don’t have access to its HTML to apply Tailwind classes directly. 
And yet I would like to use Tailwind utilities instead of writing manual CSS.
What should I do in this case if @apply is discouraged?

**Reply by wongjn** [reply]
Write manual CSS 😄

**Reply by alexchexes** [reply]
I appreciate your funny mood, but the question is serious — is there another way to accomplish this task without using `@apply` while still using Tailwind CSS utilities?

**Reply by wongjn** [reply]
No. The recommended way is to write "vanilla" CSS, referencing theme tokens via `var()` and using any complex variants via `@variant`.

**Reply by alexchexes** [reply]
So, is the recommendation now to **remember theme tokens** in addition to remembering Tailwind utilities? I don’t think Tailwind would have become so popular if it had been like that from the beginning. We loved it because it _just worked_ and made complex tasks really easy. But from what you're saying, it feels like Tailwind is abandoning this approach. Or am I wrong

**Reply by wongjn** [reply]
From my understanding, it has always been discouraged to use `@apply` (the Tweets linked above even pre-date v4).

**Reply by alexchexes** [reply]
I just checked the docs again (v4), and it looks like `@apply` exists specifically for the purpose I described above:

> This is useful when you need to write custom CSS (**like to override the styles in a third-party library**) but still want to work with your design tokens and use the same syntax you’re used to using in your HTML.

So, it's strange to read tweets with statements like:

> _if I started Tailwind CSS over from scratch, there would be no @​apply 😬_

From my perspective, _"override the styles in a third-party library"_ is a perfectly valid and commonly used case. So why wouldn’t we be able to do this using the utilities we're already used to? Why should we have to learn internals like theme variables, which lack autocompletion, change from version to version, and basically take us back to plain CSS instead of Tailwind? This idea feels insane to me

**Reply by JaquesBotha** [reply]
Exactly the point I'm trying to make. Especially working with complex backend-end UI where dynamic components are used. For example I had to deal with a dynamic quickgrid native component and the only way to get it styled in was with @apply. Also, you cannot argue that @apply should be avoided and variables should be used... Thats extremely messy and almost tripple the amount of coding. 

Clearly 4.0 is not ready yet! Still sticking with 3 untill this is sorted out.

**Reply by adamwathan** [reply]
> So, is the recommendation now to remember theme tokens in addition to remembering Tailwind utilities? 

Just use `@reference`? That's what it's for.

You know how you have to do this in a Vue file when you want to use the `ref` function?

```html
<script setup>
import { ref } from 'vue'
const greeting = ref('Hello World!')
</script>
```

...and how if you dont `import { ref } from 'vue'` you get an error that `ref` isn't defined? This is the same thing.

**Reply by alexchexes** [reply]
> Just use `@reference`?  

Thank you, Adam. Given that this advice comes from the creator of Tailwind, we understand that this is the recommended approach.  

I wasn’t complaining about the need to use `@reference`; it was just a situation where a previous commenter was insisting that we shouldn’t use `@apply` at all—even in cases that I later found explicitly mentioned in the docs as valid uses. Initially, I was just asking what the recommended approaches were. That’s it. Thanks.

**Comment by JaquesBotha** [comment]
I opened a bug report for this. Honestly love debating but I dont have time and need a solution... 

[@applyBroken in Tailwind CSS v4.0 – No Clear Fix or Docs!](https://github.com/tailwindlabs/tailwindcss/issues/16346)

**Comment by philipheinser** [comment]
`Error: Cannot apply unknown utility class: text-darkGrey`

I looked into the upgrade guide but did not find what broke it

**Comment by trustum123** [comment]
```
<style lang="postcss">
@import "tailwindcss";

.my-element {
  .div  {
    font-size: 50px;
    @apply text-amber-300;
  }
}
</style>
```

is it terrible to link to a global css file or just directly import into each tailwindcss component? Can I do this? I also encountered this problem in the next section 3.

**Reply by wongjn** [reply]
This will cause duplication in your CSS. In this situation, you'd use `@reference`:

```svelte
<style lang="postcss">
@reference "../../path/to/your/styles.css";

.my-element {
  .div  {
    font-size: 50px;
    @apply text-amber-300;
  }
}
</style>
```

**Reply by ManosKokos** [reply]
> This will cause duplication in your CSS. In this situation, you'd use `@reference`:
> 
> ```svelte
> <style lang="postcss">
> @reference "../../path/to/your/styles.css";
> 
> .my-element {
>   .div  {
>     font-size: 50px;
>     @apply text-amber-300;
>   }
> }
> </style>
> ```

I am facing similar issues because I wanted to use daisyUI 5 which uses tw 4.
I really do not like adding the `@reference "../../path/to/your/styles.css"; `
especially when the relative directory is different for various componets but to make things a little better 
I added an alias for that file in `svelte.config.js` and use that :  `@reference "$mainCSS";`
```
alias: {
  '$mainCSS': 'src/main.css',
  ...
}
```
thought it was a good idea to leave this here.

**Comment by lk77** [comment]
I have a similar issue,

"@apply xx !important" does not work anymore,

example :

````css
.icon-right {
    @apply ml-1 !important
}
````

**Reply by veslorens** [reply]
@lk77 You can fix this by adding exclamation mark at the end of every class, for example:
```
.icon-right {
    @apply ml-1!
}

.foo {
    @apply bg-light-800! text-dark-200! p-4!
}
```

**Comment by YutoYasunaga** [comment]
I'm using Rails, and got very strange error like this when build assets for production:

`Error: Cannot apply unknown utility class: ml-1`

Although `ml-1` is built-in class. It happen when I upgraded to frustrating V4.

```
@import "tailwindcss" important;
@import 'flowbite';
@import url("https://fonts.googleapis.com/css2?family=Inter:ital,opsz,wght@0,14..32,100..900;1,14..32,100..900&display=swap");
@theme {
    --color-primary-light: #047857;
    --color-primary: #065f46;
    --color-primary-dark: #064e3b;
}
@custom-variant dark (&:where(.dark, .dark *));

.svg {
    @apply ml-1;
}
```

**Comment by wu0407** [comment]
https://github.com/tailwindlabs/tailwindcss/discussions/16429
